### DISCUSSION TITLE
Add support for M29F160 - newly manufactured NOR Flash?

### DISCUSSION THREAD

**Discussion by MouseBiteLabs** [sanni/cartreader#963] [Ideas & Suggestions]
This is specifically for flashing SNES cartridges.

The M29F160 is a commercially available, newly manufactured 5V tolerant NOR Flash chip similar to the 29F016. The difference is the 160 has a 16-byte mode or an 8-byte mode, similar to the 27C160 UV EPROM chips.

I have made Game Boy cartridges that utilize these chips, and they appear to be flashable via the OSCR (under the CFI option). But, looking at the Arduino code, it doesn't appear that these chips are supported for SNES games.

Would it be possible to include support for them, to give people an option for carts made with brand new 5V tolerant components?

**Comment by sanni** [comment]
I have just added the CFI option to the SNES repro menu. I don't own a M29F160 to test it but a similar 29LV gets flashed successfully.

**Comment by MouseBiteLabs** [comment]
Wow, thank you! It works! I really appreciate it!

I do have one other question.. the 29F160s are only 16Mbit, so I also made a SNES design that uses two of them in parallel, with a 74HCT139 decoder to select between the lower 16Mbit and upper 16Mbit based on the A21 signal. I tried using the CFI flash option for this configuration, and it freezes on "Writing flash".

Is there any way to also add an option to program two of these for larger games?

**Comment by sanni** [comment]
Can't test it myself since I don't have any two flashrom repros that are supported by the CFI code but have a try:
[snes_flash_test.zip](https://github.com/user-attachments/files/15837231/snes_flash_test.zip)

If it works I'll upload it to the master branch.

**Comment by MouseBiteLabs** [comment]
The 1x mode works just as it did before, but 2x causes it to hang at "Writing flash ["
I used the same ROM file as I did with 1x mode if that makes a difference.

**Reply by sanni** [reply]
> but 2x causes it to hang at "Writing flash ["

Try changing flash.ino lines 870, 879, 974 and 983 from "PORTL ^= (1 << PL6);" to "PORTL ^= (1 << PL5);".
So it toggles SNES BA5 instead of BA6.

**Reply by MouseBiteLabs** [reply]
Ok, thanks! I won't be able to test until next week but I will update you.

**Reply by MouseBiteLabs** [reply]
Ok, I tried this fix, and it doesn't freeze anymore! However, when it tries to verify, it fails (2935299 bytes did not verify).

For reference, I dumped my copy of Super Metroid (3MB total, LoROM game) and tried to flash it to this cart with 2x 2MB chips using the CFI LoROM 2x 2MB option.

When I dumped the cart as 4MB LoROM 256K Save, I compared the file to the original ROM file, the HEX values match up to address 0x00280000 in the ROM file. I also viewed the dumped HEX file in ASCII (and byte swapped) I found this at 0x00200000, looks like it took the .txt file in the same folder and wrote it to the chip. (Everything after this is blank)

![image](https://github.com/sanni/cartreader/assets/97127539/17d4231d-b6ca-43d7-8a71-08121c175030)

**Reply by sanni** [reply]
Last version was hard coded for 4MB roms, this one is based on file size so 3MB roms should work too:
[snes_flash_test_2.zip](https://github.com/user-attachments/files/15952288/snes_flash_test_2.zip)

Maybe that fixes writing the txt file instead of the rom file.
I also added a little bit more text output to make it more clear what the OSCR is doing.

**Reply by MouseBiteLabs** [reply]
I appreciate you taking the time to look into this, by the way!

Unfortunately that still doesn't fix the issue. The text file is still written at 0x00200000, and it's completely blank afterwards. The dumped ROM also doesn't have *any* data before 0x00100000. The display reported 2-million something unverified bytes (I didn't catch the number).

If it helps: I did a search for the dumped ROM file's first 100 bytes (starting at 0x00100000, after the chunk of blank space at the beginning), and it matched at Metroid's ROM at 0x00200000.

I did a search for Metroid's first 100 bytes (starting at 0x00000000), and it did not show up anywhere in the dumped ROM file.

I tried expanding the ROM to 4MB instead and that reported 3.9 million unverified bytes. This one similarly started the ROM file at 0x00100000, and *still* has the text file included at 0x00200000 (and is blank afterwards). I'm not sure where the text file is coming from because I deleted it from the folder with the ROM in it, but it says:

```
ROM SAVE
Rom Size: 3 MB (96 banks)
Save Size: 8 KB
Checksum: F8DF
Press Button...

[+] Read ROM


Saving to SNES/ROM/Super Metroid/34/...
[*******************]
Checksum... F8DF -> OK
CRC32... D63ED5F8 -> Super Metroid (Japan, USA) (En,Ja).sfc
: ROM SAVE
Rom Size: 3 MB (96 banks)
Save Size: 8 KB
Checksum: F8DF
Press Button...

[+] Read ROM


Saving to SNES/ROM/Super Metroid/34/...
[*******************]
Checksum... F8DF -> OK
CRC32... D63ED5F8 -> Super Metroid (Japan, USA) (En,Ja).s
```

So it looks like what's happening is nothing is written for the first quarter of the ROM space, then the second half of the Metroid ROM is being written at 0x00100000, and once it hits 0x00200000, the text file is written to the chip and stops writing any more ROM data.

**Reply by sanni** [reply]
Try this version:
[snes_flash_test_3.zip](https://github.com/user-attachments/files/15959591/snes_flash_test_3.zip)
This should fix the first quarter thing.

**Reply by MouseBiteLabs** [reply]
It indeed does fix it! I'm able to flash 4MB LoROM games now with 2x 2MB. This is amazing.

However - I tried HiROM now (using DKC), and it hangs at "Writing Flash [" like LoROM did a few weeks ago. Hopefully this is an easy fix?

**Reply by sanni** [reply]
Does it hang on writing the first or the second chip?
If it's the second we're probably not hitting the right address line again to make the switch to the second flash chip.
Try this one:
[FLASH.zip](https://github.com/user-attachments/files/15968806/FLASH.zip)

**Reply by MouseBiteLabs** [reply]
It was hanging on the first chip. But that code fixed the freezing issue. Now, it programs all the way, but when I dump the ROM after flashing the cart, it's the second half of the original DKC ROM written twice.

**Reply by sanni** [reply]
Can you post a schematic of the PCB? Especially how the 74HCT139 is connected.

**Reply by MouseBiteLabs** [reply]
Ok - I made a mistake. When I put the board in HiROM mode I forgot to switch the upper address line going to the '139 decoder from BA6 to BA5 on the cart edge. I fixed this, and went back to snes_flash_test_3 files. I tried flashing DKC again, and it successfully flashes, but when it verifies it claims there are 2.2m non-verified bytes. But the ROM is still there, I can dump it and it matches the original, and it plays on the SNES. So it seems the only lingering error from that set of files is it doesn't properly verify what I'm guessing is half of the ROM even though the data was written.

(I tried using the newer Flash.ino you attached in the most recent post with the BA5 adjustment, and it also threw 2m-ish byte errors, however this time the ROM was corrupted that it wasn't readable by the OSCR. I didn't dump the file to see what the corruption was though.)

I appreciate the time you've taken to get this working. When I'm done with the design, I'll be releasing the files open source, and also buying a bunch of blank boards. I'd gladly send you some if you'd like to have them.

**Reply by sanni** [reply]
I updated the verify function to work with the 2x2MB selection and uploaded all to the [master branch](https://github.com/sanni/cartreader/archive/refs/heads/master.zip)
Hopefully we managed to squash all the bugs this time. 😄

**Reply by MouseBiteLabs** [reply]
Hmmm I'm still getting 227889 bytes to not verify (less than before!) but the ROM dump still matches the original ROM file??

**Reply by sanni** [reply]
Found two more bugs concerning ROMs smaller than 4MB. Maybe now 😄

**Reply by MouseBiteLabs** [reply]
It still fails :/
This is with the DKC ROM, which is a full 4MB

**Reply by sanni** [reply]
Could it be the SRAM that is activated on accident?
With exhirom I have this line: 
````
// Switch SNES BA6(PL6) to HIGH to disable SRAM
PORTL |= (1 << 6);
````

Maybe adding it between line 1025 and 1026 in flash.ino helps. 
https://github.com/sanni/cartreader/blob/master/Cart_Reader/FLASH.ino#L1025

**Reply by MouseBiteLabs** [reply]
For HiROM, if BA6 is low and BA5 is high, the SRAM will be active. So that does seem like what's happening right after BA5 is flipped, if BA6 is still low!

I put that code in where you specify, and it fixed it. That would also explain why the RAM was getting corrupted when I reflashed the game, but I thought maybe that was something I did haha.

Thanks again!

**Comment by Marv17** [comment]
hello , i tested your modifier snes module with CFI flash , and with a m29w320 flashchip adaptator who normally support CFI i get a CFI query failed^^

these type of chips m29w320 and m29w640 are supported to flash in game boy mode with CFI..

edit : 

it worked on an old protype with sram + 74hc139 , and not on the no save one ^^
![IMG_1812](https://github.com/sanni/cartreader/assets/132087769/c65aa1fb-9675-43c5-9b1f-79307c6a0e8d)
![IMG_1814](https://github.com/sanni/cartreader/assets/132087769/a5de44e5-79c8-43c4-84ca-c1195ada795c)

**Comment by MouseBiteLabs** [comment]
Hey @sanni -- would it be possible to add support for this for 4x 2MB chips for ExHiROM and ExLoROM as well?

**Reply by sanni** [reply]
You can play around with this, maybe you can get it to work: 
https://github.com/sanni/cartreader/archive/refs/heads/master.zip

**Reply by MouseBiteLabs** [reply]
Appreciate it! I will mess around with it this weekend, I won't have time until then.

Right now I have the ability to program the bottom two and top two independently using a switch, so I was able to successfully flash the ROM split in two using HiROM/LoROM programming. So I know the board is capable of it.

Thanks again! I'll let you know how it goes.

**Comment by StanMOI** [comment]
Tests CFI with 1x 8MB done. Flash S29GL064 and 29LV640
Not working for me.
Erasing, Writing, Verifying done successfully but on SNES I get black screen for both functions (standard and reversed).

![IMG_7826](https://github.com/user-attachments/assets/0067848b-8725-435d-aa97-c4cfdfc56cf6)
![IMG_7823](https://github.com/user-attachments/assets/e40c8cc3-d661-49f7-a988-66207693d331)

![IMG_7825](https://github.com/user-attachments/assets/c328be09-4b63-48a1-8bee-7da49d12e6f3)
![IMG_7827](https://github.com/user-attachments/assets/a76bde02-8d8e-4976-8f09-9aab18a77fa9)

**Reply by sanni** [reply]
I have updated the github code. I think I fixed the issue.

**Reply by StanMOI** [reply]
Flash S29GL064
Rom File = 8MB

**CFI LoROM / 1x 8MB** 

- standard (ABCD) = Erasing / Writing / Verifying = OK; SNES test = KO

- reversed (CDAB) = Erasing / Writing / Verifying / SNES test = ALL OK

//////////////////////////////////
Flash S29GL064
Rom File = 8MB
Rom File = 6MB

**CFI HiROM / 1x 8MB** 

If 8MB rom:
- standard (ABCD) = File too big
- standard (ABCD) = File too big

If 6MB rom:
- reversed (ABCD) = Erasing = OK / Writing = KO (Freezing)
- reversed (CDAB) = Erasing = OK / Writing = KO (Freezing)

**Reply by sanni** [reply]
I think I found one more bug with hirom flashing concerning files over 4MB, hopefully that should fix the 6MB issue.

**Reply by StanMOI** [reply]
CFI HiROM / 1x 8MB

If 8MB rom:

standard (ABCD) = Erasing = OK / Writing = KO (Freezing at the middle)
reversed (CDAB) = Erasing = OK / Writing = KO (Freezing at the middle)

If 6MB rom:

standard (ABCD) = Erasing = OK / Writing = KO (Freezing at the middle)
reversed (CDAB) = Erasing = OK / Writing = KO (Freezing at the middle)

**Reply by sanni** [reply]
Maybe now? 🐛

**Reply by StanMOI** [reply]
CFI HiROM / 1x 8MB

If 8MB rom:

standard (ABCD) = Erasing  / Writing / Verifying = OK;  SNES test = KO
reversed (CDAB) = Erasing  / Writing / Verifying / SNES test = ALL OK

If 6MB rom:

standard (ABCD) = Erasing  / Writing / Verifying = OK;  SNES test = KO
reversed (CDAB) = Erasing  / Writing / Verifying / SNES test = ALL OK

**Comment by StanMOI** [comment]
Thank you Sanni.
I think that it's will be good to add "confirmation" button, after selecting the file and just before "writing".

**Comment by StanMOI** [comment]
One more thing Sanni.
I tried to do "Flashrom Programmer/CFI" 
It's Freezing when writing 8MB file. 
And also Freezing with function "Flashrom Programmer/8bit Flash" 
I didn't tried with other size.
I used the same PCB like for CFI HiROM / 1x 8MB

**Reply by sanni** [reply]
The flashrom programmer option is for single flash chips on one of the flash adapter PCBs not for SNES repros.
The flashrom programmer option does not do any SNES lorom/exhirom/exlorom mapping and does not disable sram.

**Reply by StanMOI** [reply]
Ok 👌🏻 
Thanks

**Comment by MouseBiteLabs** [comment]
Hey @sanni! I was able to test out the latest code in the repo, and it seems to be working well, at least on my test with 4x 2MB chips for CFI HiROM. I plan to try LoROM tomorrow.

However, one thing I noticed - when I try to flash CFI HiROM with just 2x 2MB, it now hangs at the midpoint of the first 2MB chip. Do you have any idea why? I can't get it to work normally any more. Since 4x 2MB works fine and operates in a SNES, it doesn't seem like it should be an issue with the cart hardware.

For what it might be worth - the code you posted originally a few days ago in the zip file also seemed to work for me for 4x 2MB, though I believe the 2x 2MB freezing issue still existed. I'm not sure how I didn't encounter the bugs you fixed over the past few days, if that means anything!

**Reply by sanni** [reply]
Found the freeze issue, now the main thing to test would be if the flashed carts run in a SNES or if the OSCR flashes the chips in the wrong order.

**Reply by StanMOI** [reply]
CFI HiROM / 1x 8MB
8MB rom:
AB[C]D = Erasing / Writing / Verifying / SNES test = ALL OK
/////////////////////////////////////////////////////////////////////////////////////////////////////////
CFI LoROM / 1x 8MB
8MB rom:
AB[C]D = Erasing / Writing / Verifying / SNES test = ALL OK

**Reply by MouseBiteLabs** [reply]
> Found the freeze issue, now the main thing to test would be if the flashed carts run in a SNES or if the OSCR flashes the chips in the wrong order.

Awesome, everything seems to be working, at least as far as I have tested. I did both LoROM and HiROM games, 2x 2MB, with save files and they ran on the SNES. 4x 2MB HiROM (Tales of Phantasia) worked as well. Even tried an "ExLoROM" game (Thracia 776 translated) with 4x 2MB and that worked as well.

Thanks again for all the support!
